### PR TITLE
Inline Assembly

### PR DESCRIPTION
**Pull request by chriseth**
DEPENDS:
{
"alethzero": "assembly",
"mix": "assembly"
}

### CODE REVIEW

**Comment by immesys** [comment]
I can't wait! Inline assembly will make my life a ton easier. :+1:

**Comment by LianaHus** [comment]
so far so good. ping me when the PR is ready

**Comment by LianaHus** [comment]
+

**Comment by chriseth** [comment]
Yay! All checks passed! :-)

**Comment by bobsummerwill** [comment]
So ready to go, you think,@chriseth?

**Comment by chriseth** [comment]
Still need someone to review

**Comment by bobsummerwill** [comment]
So I've just looked through all the changes, but I had no insight into Solidity implementation yet, so cannot really offer any useful feedback, other than to say that I spotted no typos (well done!) and that there is surprisingly little code considering the "size" of the feature!

File: libsolidity/inlineasm/AsmCodeGen.cpp [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		boost::apply_visitor(*this, *_varDecl.value);
+		expectDeposit(1, height);
+		m_state.variables.push_back(make_pair(_varDecl.name, height));
+	}
+	void operator()(assembly::Block const& _block)
+	{
+		size_t numVariables = m_state.variables.size();
+		std::for_each(_block.statements.begin(), _block.statements.end(), boost::apply_visitor(*this));
+		// pop variables
+		//@TODO check height before and after
```
Comment (chriseth) [inline libsolidity/inlineasm/AsmCodeGen.cpp:196]: I opted for not checking the height on purpose - we could make that a bit clearer in the documentation.

**Comment by VoR0220** [comment]
Reviewing with my knowledge of the solc compiler.

**Comment by VoR0220** [comment]
For consistency, can we rename the test file in libsolidity to SolidityInlineAssembly?

File: docs/control-structures.rst [hunk 2]
```diff
@@ -167,3 +167,397 @@ Currently, there are three situations, where exceptions happen automatically in
...
+				// new "memory end" including padding
+				mstore(0x40, add(o_code, and(add(add(size, 0x20), 0x1f), bnot(0x1f))))
+				// store length in memory
+				mstore(o_code, size)
+				// actually retrieve the code
+				extcodecopy(_addr, add(o_code, 0x20), 0, size)
+			}
+		}
+	}
+
```
Comment (VoR0220) [inline docs/control-structures.rst:219]: Perhaps something here in the docs that could show a side by side comparison of how you would do it "the normal way" and then show the estimated gas used....something that could really show the benefits of utilizing this.

File: solc/CommandLineInterface.cpp [hunk 3]
```diff
@@ -574,6 +585,7 @@ bool CommandLineInterface::processInput()
 	};
 
 	m_compiler.reset(new CompilerStack(m_args.count(g_argAddStandard) > 0, fileReader));
+	auto scannerFromSourceName = [&](string const& _sourceName) -> solidity::Scanner const& { return m_compiler->scanner(_sourceName); };
```
Comment (VoR0220) [inline solc/CommandLineInterface.cpp:588]: Can I get a note here about what exactly this is doing (the syntax is a little strange here)

Comment (LianaHus) [inline solc/CommandLineInterface.cpp:588]: correct me if I'm wrong Chris, the lambda function captures objects by copy, they can't be modified inside(->), and returns const& to scanner object based on _sourceName.

File: solc/CommandLineInterface.cpp [hunk 4]
```diff
@@ -593,7 +605,8 @@ bool CommandLineInterface::processInput()
 			SourceReferenceFormatter::printExceptionInformation(
 				cerr,
 				*error,
-				(error->type() == Error::Type::Warning) ? "Warning" : "Error", *m_compiler
+				(error->type() == Error::Type::Warning) ? "Warning" : "Error",
+				scannerFromSourceName
```
Comment (LianaHus) [inline solc/CommandLineInterface.cpp:609]: sorry wrong style here

File: libsolidity/inlineasm/AsmCodeGen.cpp [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+	}
+
+private:
+	GeneratorState& m_state;
+};
+
+class CodeTransform: public boost::static_visitor<>
+{
+public:
+	/// Create the code transformer whit appends assembly to _state.assembly when called
```
Comment (LianaHus) [inline libsolidity/inlineasm/AsmCodeGen.cpp]: typo whit

File: libsolidity/inlineasm/AsmParser.cpp [hunk 6]
```diff
@@ -78,28 +80,28 @@ AsmData::Statement InlineAssemblyParser::parseStatement()
...
 		if (m_scanner->currentToken() == Token::Assign)
 		{
 			// functional assignment
 			m_scanner->next();
-			unique_ptr<AsmData::Statement> value;
-			value.reset(new AsmData::Statement(parseExpression()));
-			return AsmData::FunctionalAssignment{{move(name)}, move(value)};
+			unique_ptr<Statement> value;
+			value.reset(new Statement(parseExpression()));
+			return FunctionalAssignment{{std::move(name)}, std::move(value)};
```
Comment (LianaHus) [inline libsolidity/inlineasm/AsmParser.cpp:100]: is there any special reason you have "std::"s here?

File: solc/CommandLineInterface.cpp (hunk 3 above)
Comment (chriseth) [inline solc/CommandLineInterface.cpp:588]: This is c++11 syntax for lambda functions. The `[&]` captures by reference and not by copy (`[=]` would be copy). So `scannerFromSourceName` is a function object that returns the scanner that belongs to the given source name.

File: libsolidity/inlineasm/AsmParser.cpp (hunk 6 above)
Comment (chriseth) [inline libsolidity/inlineasm/AsmParser.cpp:100]: Yes, some compiler complained about ambiguous identifiers. I think there is also some `boost::move`.

File: docs/control-structures.rst (hunk 2 above)
Comment (chriseth) [inline docs/control-structures.rst:219]: Will add a note that explains that this is not possible at all without inline assembly.
Will also add another example that shows how to save gas using inline assembly.

File: solc/CommandLineInterface.cpp (hunk 4 above)
Comment (chriseth) [inline solc/CommandLineInterface.cpp:609]: Wrong style? What do you mean?

**Comment by VoR0220** [comment]
#459

**Comment by chriseth** [comment]
@VoR0220 updated the documentation, can you check please?

File: solc/CommandLineInterface.cpp (hunk 4 above)
Comment (LianaHus) [inline solc/CommandLineInterface.cpp:609]: either scannerFromSourceName should go up or "Error" should have own line

Comment (LianaHus) [inline solc/CommandLineInterface.cpp:609]: sorry you are right, ignore my comments

**Pull request merged**